### PR TITLE
Replace windows convert case functions

### PR DESCRIPTION
**Pull request by galanCA**
## Description

Updates windows functions that change the case of a string either upper (strupr) or lower (strlwr). As recommended by windows. It also creates a shadow function used in GCC and Clang for these functions.

### Test

A unit test is added to this PR to ensure their behavior stays the same.

## Author Progress Checklist:

- [ ] Open draft pull request
    - [ ] Make title clearly understandable in a standalone change log context
    - [ ] Assign yourself the issue
    - [ ] Add at least one label (enhancement, bug, or maintenance)
    - [ ] Link the issue(s) addressed by this PR (under "Development" in the sidebar menu) 
- [ ] Make code changes (if you haven't already)
- [ ] Self-review of code
    - [ ] My code follows the style guidelines of this project
    - [ ] I have added comments to my code, particularly in hard-to-understand areas
    - [ ] I have only committed the necessary changes for this fix or feature
    - [ ] I have made corresponding changes to the documentation
    - [ ] My changes generate no new warnings
    - [ ] I have ensured that my fix is effective or that my feature works as intended by:
        - [ ] exercising the code changes in the test framework, and/or
        - [ ] manually verifying the changes (as explained in the the pull request description above)
    - [ ] My changes pass all local tests
    - [ ] My changes successfully passes CI checks
    - [ ] Add any unit test for proof and documentation.
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
- [ ] Move pull request out of draft mode and assign reviewers
- [ ] Iterate with reviewers until all changes are approved
    - [ ] Make changes in response to reviewer comments
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
    - [ ] Re-request review in GitHub

## Reviewer Checklist:

 - [ ] Read the pull request description
 - [ ] Perform a code review on GitHub
 - [ ] Confirm all CI checks pass and there are no build warnings
 - [ ] Pull, build, and run automated tests locally
 - [ ] Perform manual tests of the fix or feature locally
 - [ ] Add any review comments, if applicable
 - [ ] Submit review in GitHub as either
     - [ ] Request changes, or
     - [ ] Approve
 - [ ] Iterate with author until all changes are approved

### CODE REVIEW

**Review by chipbarnaby — CHANGES_REQUESTED** [review]
I don't think this is the right approach, especially compareCaseStr().  We do not want to pay the price of constructing std::strings just to be able to compare them.  That often involves memory allocation.

I believe there are valid standard functions for all the null terminated string operations we want to do.  Actually, is it known that the current functions are not available via libraries on other platforms?  I believe we are using standard C functions that are generally widely supported.  In other words, why is this necessary at all (or so broadly necessary).

At some point we may convert to using std::string, at which point std::string operations can be used without unnecessary conversion.

**Comment by galanCA** [comment]
I will implement a similar system as PR #256.

**Comment by galanCA** [comment]
I will implement a similar system as PR #256.

File: test/unit/strpak.unit.cpp [hunk 1]
```diff
@@ -28,3 +28,25 @@ TEST(strpak, compare_case_insensitive_string)
 	EXPECT_LT( _strnicmp(long_string_upper_case, long_string_wrong_case, 22), 0);
 	EXPECT_EQ( _strnicmp(long_string_upper_case, long_string_wrong_case, 10), 0);
 }
+
+TEST(strpak, convert_case_functions)
+{
+	// Set up strings
+	char uppercaseString[] = "this should be uppercase.";
+	char lowercaseString[] = "THIS SHOULD BE LOWERCASE.";
+	char randomCaseString[] = "Hello There ARE many case HERE!";
```
Comment (tanaya-bigladder) [inline test/unit/strpak.unit.cpp:37]: Perhaps the random case string could also put an upper-case letter in the middle of a word.

Comment (tanaya-bigladder) [inline test/unit/strpak.unit.cpp:37]: Also, "there are many case**s**" here. :)

File: test/unit/strpak.unit.cpp [hunk 2]
```diff
@@ -28,3 +28,25 @@ TEST(strpak, compare_case_insensitive_string)
...
+	char numberString[] = "There are 12311 numbers (#).";
```
Comment (tanaya-bigladder) [inline test/unit/strpak.unit.cpp:38]: Should we also test number/letter mixed words, as you sometimes see in variable names, e.g. "ASHRAE205"

**Review by tanaya-bigladder — CHANGES_REQUESTED** [review]

File: test/unit/strpak.unit.cpp [hunk 3]
```diff
@@ -28,3 +28,25 @@ TEST(strpak, compare_case_insensitive_string)
...
+
+	// Test convert uppercase strings
+	EXPECT_STREQ( _strupr(upper_case_string), "THIS SHOULD BE UPPERCASE.");
+	EXPECT_STREQ( _strupr(mixed_case_string), "HELLO THERE ARE MANY CASES HERE!");
+	EXPECT_STREQ( _strupr(number_string), "TESTING NUMBERS: ASHRAE205, 1252345 (#).");
+
+	
+	// Test convert lowercase strings
+	strcpy(mixed_case_string, "Hello There ARE maNy cases HERE!");
+	strcpy(number_string, "Testing numbers: ASHRAE205, 1252345 (#).");
```
Comment (tanaya-bigladder) [inline test/unit/strpak.unit.cpp:48]: I would create a new char array here; the reason is that if someone wanted to update this function later on, they could mistakenly strcpy a longer string into an already allocated char array and have a memory leak.

**Review by tanaya-bigladder — DISMISSED** [review]

**Review by chipbarnaby — CHANGES_REQUESTED** [review]
There could be trouble here due to signed characters.  tolower / toupper take integers.  That means a char > 127 will become a negative value.  The standard work-around is the following or words to that effect.

char* _strlwr(char* s)
{
    char* tmp = s;

    for (;*tmp;++tmp) {
        *tmp = tolower((unsigned char) *tmp);
    }

    return s;
}


This character stuff has always been tricky, but I think defining _strupr and _strlwr is a good enough solution pending conversion to std::string.

I would flag these functions as ?MULTIPLATFORM? -- this will work but certainly not elegant.

**Review by chipbarnaby — APPROVED** [review]

**Review by nealkruis — APPROVED** [review]

**Comment by nealkruis** [comment]
@tanaya-bigladder merging is currently blocked because of your "request changes". Can you re-review?

**Review by tanaya-bigladder — COMMENTED** [review]
Approved; sorry it got out of sync.

**Pull request merged**